packages/router/router.ts

```diff
@@ -3912,15 +3972,88 @@ async function loadLazyRouteModule(
     ...mapRouteProperties(routeToUpdate),
     lazy: undefined,
   });
+
+  return route;
 }
 
-async function callLoaderOrAction(
+function createCallLoaderOrAction(dataStrategy: DataStrategyFunction) {
+  return async (
+    type: "loader" | "action",
+    request: Request,
+    match: AgnosticDataRouteMatch,
+    matches: AgnosticDataRouteMatch[],
+    manifest: RouteManifest,
+    mapRouteProperties: MapRoutePropertiesFunction,
+    basename: string,
+    v7_relativeSplatPath: boolean,
+    opts: {
+      isStaticRequest?: boolean;
+      isRouteRequest?: boolean;
+      requestContext?: unknown;
+    } = {}
+  ): Promise<DataResult> => {
+    let [result] = await dataStrategy({
+      matches: [
+        finesseToAgnosticDataStrategyMatch(match, mapRouteProperties, manifest),
+      ],
+      request,
+      type,
+      defaultStrategy(match) {
+        return callLoaderOrActionImplementation(
+          type,
+          request,
+          match,
+          matches,
+          basename,
+          v7_relativeSplatPath,
+          opts
+        );
+      },
+    });
+
+    return result;
+  };
+}
+
+function finesseToAgnosticDataStrategyMatch(
+  match: AgnosticDataRouteMatch,
+  mapRouteProperties: MapRoutePropertiesFunction,
+  manifest: RouteManifest
+): AgnosticDataStrategyMatch {
+  let loadRoutePromise: Promise<AgnosticDataRouteObject> | undefined;
+
+  if (match.route.lazy) {
+    try {
+      loadRoutePromise = loadLazyRouteModule(
+        match.route,
+        mapRouteProperties,
+        manifest
+      );
+    } catch (error) {
+      loadRoutePromise = Promise.reject(error);
+    }
+  }
+
+  if (!loadRoutePromise) {
+    loadRoutePromise = Promise.resolve(match.route);
+  }
+
+  loadRoutePromise.catch(() => {});
+
+  return {
+    ...match,
+    route: Object.assign(
+      loadRoutePromise,
+      match.route
+    ) as unknown as LazyRoutePromise,
```

Comment:
This type is sort of inherently messy 😕 .  This is "right" since we want the statically defined fields available without awaiting the promise (like `id`/`index`/`path`/`children`/etc and possibly `loader`/`action`).  So the legit type is something along the lines of this I guess?

```ts
route: Promise<AgnosticDataRouteObject> & {
  [k: keyof ImmutableRouteKey]?: AgnosticRouteObject[k],
  loader?: LoaderFunction
  action?: ActionFunction
}
```

Unsure of a better alternative offhand though...unless we left `match.route` alone and added a sibling `match.lazyRoute` that was the promise and only got populated if `route.lazy` existed...